### PR TITLE
[WIP] Auto set trace-agent cpu limit based on any k8s limit

### PR DESCRIPTION
**Pull request by ajgajg1134**
**This PR should wait until https://github.com/DataDog/helm-charts/pull/794 is merged to avoid a situation where new installs of the agent get the default 50% cpu limit instead of the new behavior or the old behavior**

### What does this PR do?
This PR sets the trace-agent's internal CPU limit to 80% of the configured K8S CPU limit. Note this is set in this PR here: https://github.com/DataDog/helm-charts/pull/794. Note that the Trace-Agent limit, does NOT function like the Kubernetes limit. The trace-agent limiter is a soft-limit that kicks in only AFTER the total cpu usage of the trace-agent exceeds the configured limit. 

### Motivation
Currently in K8s environments the trace-agent only relies on the container runtime / k8s for cpu throttling. This means that the trace-agent can become overwhelmed with traffic but not report this to the backend leading to customers being unaware their trace-agents need more resources (or less traffic). Using the k8s limit in the trace-agent will allow it to rate limit (read: drop) traffic when it starts to be overwhelmed, and indicate to the backend (via tags) that the trace-agent has insufficient resources.

### Possible Drawbacks / Trade-offs
This is a behavior change from the current "no limit except for k8s container limit". For customers who have very consistent traffic and sit between 80-100% of their utilization this change will cause some of their traffic to be dropped / rate limited as the trace-agent tries to get to 80% usage. However, customers can continue to see the same behavior by just disabling the cpu limiter as it was disabled in k8s by default before by setting `apm_config.max_cpu_percent` to 0.  

### Describe how to test/QA your changes
I installed the datadog agent using my branch of the helm chart to my local minikube cluster. I configured a CPU limit on the container resources and verified that the trace-agent both:
1. Logged that it was using the K8S limit
2. Began to throttle traffic once the limit was reached.
Running this same test on any RC builds should be sufficient here.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: releasenotes/notes/traceAgentK8sLimit-165e630cca46f6a0.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+enhancements:
+  - |
+    APM: Trace-Agent will now set internal CPU rate limiter based on detected Kubernetes CPU Limit. To continue ONLY using the K8S cpu limit you must set `apm_config.max_cpu_percent` or `DD_APM_MAX_CPU_PERCENT` to 0.
```
Comment (buraizu) [inline releasenotes/notes/traceAgentK8sLimit-165e630cca46f6a0.yaml:11]:
```suggestion
    APM: Trace-Agent now sets the internal CPU rate limiter based on the detected Kubernetes CPU limit. To continue using only the K8S CPU limit, you must set `apm_config.max_cpu_percent` or `DD_APM_MAX_CPU_PERCENT` to 0.
```

**Review by buraizu — CHANGES_REQUESTED** [review]

**Review by vboulineau — CHANGES_REQUESTED** [review]
I think there's a possible more straightforward, config-less approach.
We're providing a way to get self cgroup data using in `pkg/util/cgroups/self_reader.go`.

It's already used by DSD to get its own memory limit:
https://github.com/DataDog/datadog-agent/blob/main/pkg/dogstatsd/listeners/ratelimit/cgroup_memory_usage_linux.go

You could reuse it to get CPU Limit. This approach has the benefit to work everywhere on Linux (not only Kubernetes).
(Note: it does not work on Windows, we could think of a way to make it work on Windows when possible, if that's a requirement)

Also, if I may, the Agent CPU limits are normally low compared to host CPU, so using 90% to 95% of allocated CPU should not be a problem.
Hence, if the limiter is accurate enough I'd bump the the default `80%` to `90/95`

**Review by buraizu — APPROVED** [review]

File: cmd/trace-agent/config/config_linux.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
+//go:build linux
+// +build linux
+
+package config
+
+import (
+	"errors"
+
+	"github.com/DataDog/datadog-agent/pkg/util/cgroups"
+	"github.com/DataDog/datadog-agent/pkg/util/pointer"
```
Comment (vboulineau) [inline cmd/trace-agent/config/config_linux.go:10]: I'm not sure you can import `util/pointer` and `util/system` unfortunately.

File: cmd/trace-agent/config/config_linux.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	systemutils "github.com/DataDog/datadog-agent/pkg/util/system"
+)
+
+func getCgroupCPULimit() (float64, error) {
+	reader, err := cgroups.NewSelfReader("/proc", true)
```
Comment (vboulineau) [inline cmd/trace-agent/config/config_linux.go:15]: You cannot pass `true` as second parameter. Cgroups limit can also be set through `systemd` (host install)

**Pull request closed without merging**